PhysicsTools/PatUtils/plugins/L1PrefiringWeightProducer.cc

```diff
@@ -0,0 +1,478 @@
+      if (fabs(eta_jet) > 3.)
+        continue;
+      if (jetMaxMuonFraction_ > 0 && jet.muonEnergyFraction() > jetMaxMuonFraction_)
+        continue;
+      //Loop over photons to remove overlap
+      double nonprefiringprobfromoverlappingphotons = 1.;
+      for (const auto& photon : thePhotons) {
+        double pt_gam = photon.pt();
+        double eta_gam = photon.eta();
+        double phi_gam = photon.phi();
+        if (pt_gam < 20.)
+          continue;
+        if (fabs(eta_gam) < 2.)
+          continue;
+        if (fabs(eta_gam) > 3.)
+          continue;
+        double dR = reco::deltaR(eta_jet, phi_jet, eta_gam, phi_gam);
+        if (dR > 0.4)
+          continue;
+        double prefiringprob_gam = getPrefiringRate(eta_gam, pt_gam, h_prefmap_photon, fluct);
+        nonprefiringprobfromoverlappingphotons *= (1. - prefiringprob_gam);
+      }
+      //useEMpt =true if one wants to use maps parametrized vs Jet EM pt instead of pt.
+      if (useEMpt_)
+        pt_jet *= (jet.neutralEmEnergyFraction() + jet.chargedEmEnergyFraction());
+      double nonprefiringprobfromoverlappingjet = 1. - getPrefiringRate(eta_jet, pt_jet, h_prefmap_jet, fluct);
+
+      if (nonprefiringprobfromoverlappingphotons == 1.) {
+        nonPrefiringProba[fluct] *= nonprefiringprobfromoverlappingjet;
+        nonPrefiringProbaJet[fluct] *= nonprefiringprobfromoverlappingjet;
+      }
+      //If overlapping photons have a non prefiring rate larger than the jet, then replace these weights by the jet one
+      else if (nonprefiringprobfromoverlappingphotons > nonprefiringprobfromoverlappingjet) {
+        if (nonprefiringprobfromoverlappingphotons != 0.) {
+          nonPrefiringProba[fluct] *= nonprefiringprobfromoverlappingjet / nonprefiringprobfromoverlappingphotons;
+          nonPrefiringProbaJet[fluct] *= nonprefiringprobfromoverlappingjet / nonprefiringprobfromoverlappingphotons;
+        } else {
+          nonPrefiringProba[fluct] = 0.;
+          nonPrefiringProbaJet[fluct] = 0.;
+        }
+      }
+      //Last case: if overlapping photons have a non prefiring rate smaller than the jet, don't consider the jet in the event weight, and do nothing.
+    }
+    for (const auto& muon : theMuons) {
+      double pt = muon.pt();
+      double phi = muon.phi();
+      double eta = muon.eta();
+      // Remove crappy tracker muons which would not have prefired the L1 trigger
+      if (pt < 5 && !muon.isStandAloneMuon())
+        continue;
+      double prefiringprob_mu = getPrefiringRateMuon(eta, phi, pt, fluct);
+      nonPrefiringProba[fluct] *= (1. - prefiringprob_mu);
+      nonPrefiringProbaMuon[fluct] *= (1. - prefiringprob_mu);
+    }
+  }
+  auto nonPrefiringProb = std::make_unique<double>(nonPrefiringProba[0]);
+  auto nonPrefiringProbUp = std::make_unique<double>(nonPrefiringProba[1]);
+  auto nonPrefiringProbDown = std::make_unique<double>(nonPrefiringProba[2]);
+  iEvent.put(std::move(nonPrefiringProb), "nonPrefiringProb");
```

Comment:
The `unique_ptr` (with extra allocation) could be avoided with
```cpp
// in class declaration
const edm::EDPutTokenT<double> nonPrefiringProbToken_;

// in constructor
nonPrefiringProbToken_(produces<double>("nonPrefiringProb"))

// here
iEvent.emplace(nonPrefiringProbToken_, nonPrefiringProba[0]);
```

Reply:
Done